File: Persistence/src/persistence.java
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Vector;

public class persistence {

    private Vector<Simplex> F; // filtration vector. Vector is to be sorted in constructor private
    private ArrayList<ArrayList<Integer>> M; // initial boundary matrix or reduced matrix
    private int size; // F.size, number of sigma

    public persistence(String filename) throws FileNotFoundException {
        System.out.println("Reading from " + filename + "...\n");
        F = ReadFiltration.readFiltration(filename);

        System.out.println("Sorting filtration...\n");
        ReadFiltration.sortSimplex(F);
        size = F.size();

        System.out.println("Computing border matrix...");
        computeMatrix();
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++)
                System.out.print(M.get(j).get(i) + " ");
            System.out.println();
        }
        System.out.println();

    }

    private void computeMatrix() {
        int m = F.size();
        M = new ArrayList<ArrayList<Integer>>();

        for (int j = 0; j < m; j++) {
            M.add(new ArrayList<Integer>());
            for (int i = 0; i < m; i++)
                M.get(j).add(0);
        }

        for (int j = 0; j < F.size(); j++) {// F is sorted
            Simplex s = F.get(j);
            Integer[] nodes = s.vert.toArray(new Integer[s.dim + 1]);
            if (s.dim >= 1)// is simplex has dimension 0 then is border is null
                for (int rmv : nodes) {
                    s.vert.remove(rmv);// remove one of the nodes
                    for (int i = 0; i < F.size(); i++) {// F is sorted
                        Simplex b = F.get(i);
                        if (b.dim == s.dim - 1)
                            if (b.vert.equals(s.vert))
                                M.get(j).set(i, 1);
                    }
                    s.vert.add(rmv);
                }
        }
    }

    private int low(int j) {
        ArrayList<Integer> j_col = M.get(j);
        int i = M.get(j).size() - 1;

        while (i >= 0) {
            if (j_col.get(i) != 0) {
                break;
            }
            i--;
        }

        return i;
    }

    public void reduction() {
        System.out.println("Reducing matrix...");

        size = F.size();
        System.out.println(size);
        ArrayList<Integer> lowPosition = new ArrayList<Integer>(); // position of the first(low = j)
        for (int i = 0; i < size; i++) {
            lowPosition.add(-1);
        }
        for (int i = 0; i < size; i++) { // i is the indice of the column we are working on
            ArrayList<Integer> column = M.get(i);
            int low = low(i); // actual low
            if (low > -1) { // if not null column
                if (lowPosition.get(low) > -1) { // if we already have this low we add the two columns
                    while (low > -1 && lowPosition.get(low) > -1) {
                        for (int k = 0; k <= low; k++) {
                            column.set(k, (column.get(k) + M.get(lowPosition.get(low)).get(k)) % 2);
                        }
                        low = low(i);
                    }
                }
                if (low > -1) {
                    lowPosition.set(low, i);
                }
            }
        }

        // display
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++)
                System.out.print(M.get(j).get(i) + " ");
            System.out.println();
        }
        System.out.println();
    }

    Vector<Interval> computeBarCode() {
        Vector<Interval> bar_code = new Vector<Interval>();

        Vector<Integer> which_column = new Vector<Integer>(M.size());
        for (int j = 0; j < M.size(); j++) {
            which_column.add(-1);
        }

        for (int j = 0; j < M.size(); j++) {
            if (low(j) > -1) {
                which_column.set(low(j), j);
            }
        }

        // going through the columns
        for (int j = 0; j < M.size(); j++) {
            // calculating low
            int l = low(j);

            // if there's no low -> start is infinite
            if (l < 0) {
                if (which_column.get(j) == -1) {
                    bar_code.add(new Interval(F.get(j).dim, F.get(j).val));
                } else {
                    bar_code.add(new Interval(F.get(j).dim, F.get(j).val, F.get(which_column.get(j)).val));
                }
            }

        }

        for (Interval bc : bar_code) {
            System.out.println(bc);
        }

        System.out.println("\nBar-Code representation:\n");

        for (Interval bc : bar_code) {
            String op = "";
            float i = 0;
            while (i < bc.b) {
                op += " ";
                i++;
            }
            op += "|";
            if (bc.finite) {
                while (i < bc.d) {
                    op += "-";
                    i++;
                }
            } else {
                while (i < 50) {
                    op += "-";
                    i++;
                }

                op += ">";
            }
            System.out.print("dim: " + bc.k + " : t=0");
            System.out.println(op);
        }

        return bar_code;
    }

}

File: Python/MyUtils/Persistence.py
from MyUtils.Interval import Interval
from MyUtils.Simplex import Simplex
from operator import attrgetter

# Class for generating persistance diagram.

class Persistence:
    def __init__(self, FILE_PATH):
        self.M = []
        self.simplices = []
        # Reading the input.

        print(f"Reading from {FILE_PATH}...\n")
        with open(FILE_PATH) as file:
            for line in file:
                line = line.rstrip().split()
                vertices = line[2:]
                vertices = set(map(lambda x: int(x), vertices))
                self.simplices.append(
                    Simplex(
                        discoveredAt=float(line[0]),
                        dimension=int(line[1]),
                        vertices=vertices,
                    )
                )

        # Sorting on time of discovery, dimension.
        print("Sorting the filteration...")
        self.simplices = sorted(
            self.simplices, key=attrgetter("discoveredAt", "dimension")
        )
        self.num_simplices = len(self.simplices)
        # print(* self.simplices)

        print("Computing border matrix...")
        self.computeMatrix()

        # print("Reducing border matrix...")
        # self.reduceMatrix()
        # printMatrix(self.M)


    # Generating border matrix.
    def computeMatrix(self):
        m = self.num_simplices
        self.M = []

        for j in range(0, m):
            self.M.append([])
            for i in range(0, m):
                self.M[j].append(0)

        for j in range(0, m):
            s: Simplex = self.simplices[j]
            nodes = list(s.vertices)
            if s.dimension >= 1:  # If simplex has dimension 0 then is border is null
                for rmv in nodes:
                    s.vertices.remove(rmv)
                    # Remove one of the nodes
                    for i in range(0, m):  # F is sorted
                        b: Simplex = self.simplices[i]
                        if b.dimension == s.dimension - 1:
                            if b.vertices == s.vertices:
                                self.M[j][i] = 1

                    s.vertices.add(rmv)
        
        for i in range(0, self.num_simplices):
            for j in range(0, self.num_simplices):
                print(self.M[j][i], end=" ")
            print()
        
        print()

    # Gets the index of the lowest non-zero value in a column.
    def low(self, j: int) -> int:
        j_column = self.M[j]
        i = len(j_column) - 1

        while i >= 0:
            if j_column[i] != 0:
                break

            i -= 1

        return i

    # Reduction.
    def reduceMatrix(self):
        # lowPosition[i] will store the index of the lowest non-zero value in i-th column of matrix.
        lowPosition = []
        for i in range(0, self.num_simplices):
            lowPosition.append(-1)

        for i in range(0, self.num_simplices):
            i_column = self.M[i]
            low_idx = self.low(i)

            if low_idx > -1:
                if lowPosition[low_idx] > -1:
                    while low_idx > -1 and lowPosition[low_idx] > -1:
                        for j in range(0, low_idx + 1):
                            i_column[j] = (
                                i_column[j] + self.M[lowPosition[low_idx]][j]
                            ) % 2
                          
                        low_idx = self.low(i)
                if low_idx > -1:
                    lowPosition[low_idx] = i
        
        for i in range(0, self.num_simplices):
            for j in range(0, self.num_simplices):
                print(self.M[j][i], end=" ")
            print()
        print()

    # Generating bar-code intervals.
    def generateBarCodeIntervals(self) -> list:
        barCode = []

        whichColumn = []
        for kk in range(0, self.num_simplices):
            whichColumn.append(-1)

        for j in range(0, self.num_simplices):
            if (self.low(j) > -1):
                whichColumn[self.low(j)] = j
            

        # Going through the columns.
        for j in range(0, self.num_simplices):
            # Calculatin index of low value.
            l = self.low(j)

            # // If there is no low value -> start is infinite
            if (l < 0):
                if (whichColumn[j] == -1):
                    newInterval = Interval(self.simplices[j].dimension, self.simplices[j].discoveredAt, None)
                    barCode.append(newInterval)
                else:
                    newInterval = Interval(self.simplices[j].dimension, self.simplices[j].discoveredAt, self.simplices[whichColumn[j]].discoveredAt)
                    barCode.append(newInterval)
        
        return(barCode)

